File: validation.py
import re

def validate_nombre(nombre):
    if not nombre:
        return False
    return 3 <= len(nombre.strip()) <= 80

def validate_email(email):
    if not email:
        return False
    length_valid = len(email) > 15
    format_valid = re.match(r'^[\w.]+@[a-zA-Z_]+?\.[a-zA-Z]{2,3}$', email)
    return length_valid and format_valid

def validate_phone_number(phone_number):
    if not phone_number:
        return False
    length_valid = len(phone_number) >= 8
    format_valid = re.match(r'^[0-9]+$', phone_number)
    return length_valid and format_valid

def validate_select(select):
    return bool(select)

def validate_device_nombre(dnombre):
    if not dnombre:
        return False
    return 3 <= len(dnombre.strip()) <= 80

def validate_files(files):
    if not files or len(files) == 0:
        return False
    length_valid = 1 <= len(files) <= 3
    type_valid = all(
        file.mimetype.startswith("image/") or file.mimetype == "application/pdf"
        for file in files
    )
    return length_valid and type_valid

def validate_years_of_use(years_of_use):
    if not years_of_use:
        return False
    try:
        years = int(years_of_use)
        return 1 <= years <= 99
    except ValueError:
        return False

    
def validate_comment(text):
    if not text:
        return False
    return (3 < text.length and text.length < 80)

'''
def validate_form(data):
    invalid_inputs = []
    is_valid = True

    def set_invalid_input(input_nombre):
        nonlocal is_valid
        invalid_inputs.append(input_nombre)
        is_valid = False

    # Extract form data
    nombre = data.get("nombre")
    email = data.get("email")
    phone_number = data.get("phone")
    region = data.get("region")
    comuna = data.get("comuna")
    devices = data.getlist("devices")  # Assuming device data is in a list of dictionaries

    # Validate form fields
    if not validate_nombre(nombre):
        set_invalid_input("Nombre")

    if not validate_email(email):
        set_invalid_input("Email")

    if phone_number and not validate_phone_number(phone_number):
        set_invalid_input("Número")

    if not validate_select(region):
        set_invalid_input("Región")

    if not validate_select(comuna):
        set_invalid_input("Comuna")

    # Validate each device
    for index, device in enumerate(devices):
        dnombre = device.get("dnombre")
        tipo = device.get("tipo")
        years_of_use = device.get("years_of_use")
        state = device.get("state")
        files = device.get("files")

        device_number = f"Dispositivo {index + 1}"

        if not validate_device_nombre(dnombre):
            set_invalid_input(f"{device_number}: Nombre del dispositivo")

        if not validate_select(tipo):
            set_invalid_input(f"{device_number}: Tipo de dispositivo")

        if not validate_years_of_use(years_of_use):
            set_invalid_input(f"{device_number}: Años de uso")

        if not validate_select(state):
            set_invalid_input(f"{device_number}: Estado de Funcionamiento")

        if not validate_files(files):
            set_invalid_input(f"{device_number}: Fotos")

    if not is_valid:
        return False, invalid_inputs

    return True, []
'''

File: static/js/validation.js

const validatenombre = (nombre) => {
    if(!nombre) return false;
    let lengthValid = nombre.trim().length >= 3 && nombre.trim().length <= 80;
    let re2 = /^[a-zA-Z\s]+$/; //nombre sin y con espacios
    let formatValid = re2.test(nombre);
    return formatValid && lengthValid;
};
  
const validateEmail = (email) => {
    if (!email) return false;
    let lengthValid = email.length < 31;
  
    // validamos el formato
    let re = /^[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}$/;
    let formatValid = re.test(email);
  
    // devolvemos la lógica AND de las validaciones.
    return lengthValid && formatValid;
};
  
const validatecelularNumber = (celularNumber) => {
    if (!celularNumber) return false;
    // validación de longitud
    let lengthValid = celularNumber.length >= 8;
  
    // validación de formato
    let re = /^[0-9]+$/;
    let formatValid = re.test(celularNumber);
  
    // devolvemos la lógica AND de las validaciones.
    return lengthValid && formatValid;
};

const validateSelect = (select) => {
    if(!select) return false;
    return true
};

const validateDevicenombre = (dnombre) => {
  if(!dnombre) return false;
  let lengthValid = dnombre.trim().length >= 3 && dnombre.trim().length <= 80;
  return lengthValid;
};
  
const validateFiles = (archivos) => {
  if (!archivos) return false;

  // validación del número de archivos
  let lengthValid = 1 <= archivos.length && archivos.length <= 3;

  // validación del tipo de archivo
  let typeValid = true;

  for (const file of archivos) {
    // el tipo de archivo debe ser "image/<foo>" o "application/pdf"
    let fileFamily = file.type.split("/")[0];
    typeValid &&= fileFamily == "image" || file.type == "application/pdf";
  }

  // devolvemos la lógica AND de las validaciones.
  return lengthValid && typeValid;
};

const validateYearsOfUse = (yearsOfUse) => {
  if (!yearsOfUse) return false;
  let rangeValid = yearsOfUse >= 1 && yearsOfUse <= 99;
  return rangeValid;
}
  
const validateForm = () => {
  let myForm = document.forms["form-donacion"];
  let email = myForm["email"].value;
  let celularNumber = myForm["celular"].value;
  let nombre = myForm["nombre"].value;
  let region = myForm["region"].value;
  let comuna = myForm["comuna"].value;

  // variables auxiliares de validación y función.
  let invalidInputs = [];
  let isValid = true;
  const setInvalidInput = (inputnombre) => {
    invalidInputs.push(inputnombre);
    isValid &&= false;
  };
  // lógica de validación
  if (!validatenombre(nombre)) {
    setInvalidInput("Nombre");
  }
  if (!validateEmail(email)) {
    setInvalidInput("Email");
  }
  if (celularNumber && !validatecelularNumber(celularNumber)) {
    setInvalidInput("Número");
  }
  if (!validateSelect(region)) {
    setInvalidInput("Región");
  }
  if (!validateSelect(comuna)) {
    setInvalidInput("Comuna");
  }
  // Validar cada dispositivo de forma individual
  let devices = document.querySelectorAll('#deviceContainer .device'); 

  if (devices !== null) {
    devices.forEach((device, index) => {
      let dnombre = device.querySelector('input[name="dnombre[]"]').value;
      let deviceType = device.querySelector('select[name="select-device[]"]').value;
      let yearsOfUse = device.querySelector('input[name="years_of_use[]"]').value;
      let state = device.querySelector('select[name="select-state[]"]').value;
      let archivos = device.querySelector('input[name="files[]"]').files;

      // Agregar índice al nombre del dispositivo para diferenciar entre los dispositivos
      let deviceNumber = `Dispositivo ${index + 1}`;

      // Validaciones para cada dispositivo
      if (!validateDevicenombre(dnombre)) {
          setInvalidInput(`${deviceNumber}: Nombre del dispositivo`);
      }
      if (!validateSelect(deviceType)) {
          setInvalidInput(`${deviceNumber}: Tipo de dispositivo`);
      }
      if (!validateYearsOfUse(yearsOfUse)) {
          setInvalidInput(`${deviceNumber}: Años de uso`);
      }
      if (!validateSelect(state)) {
          setInvalidInput(`${deviceNumber}: Estado de Funcionamiento`);
      }
      if (!validateFiles(archivos)) {
          setInvalidInput(`${deviceNumber}: Fotos`);
      }
    });
  };
  
  // finalmente mostrar la validación
  let popup = document.getElementById("popup");
  let mensaje = document.getElementById("popup-msg");
  let listaErrores = document.getElementById("lista-popup");
  let b1 = document.getElementById("b1");
  let b2 = document.getElementById("b2");
  
  

  if (!isValid) {
    // agregar elementos inválidos al elemento val-list.
    for (input of invalidInputs) {
      let listElement = document.createElement("li");
      listElement.innerText = input;
      popup.append(listElement);
    }
    // establecer val-msg
    mensaje.innerText = "Los siguientes campos son inválidos:";

    // hacer visible el mensaje de validación
    popup.hidden = false;
  } else {
    popup.hidden = true;
    let confirmacion = document.getElementById("popup-msg");
    confirmacion.innerText = "¡Formulario válido! ¿Deseas enviarlo o volver?";
    popup.hidden = false;

    
    b1.addEventListener("click", () => popup.hidden = true );
    b2.addEventListener("click", Submit);
  }
};


Submit = () => {
  let form = document.getElementById("form-donacion");
  form.submit();
};


let submitButton = document.getElementById("submit-button");
submitButton.addEventListener("click", validateForm);